### ISSUE TITLE
Add a setting to only use IDE0180 Use tuple swap if it will avoid extra temps

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#65815]
Related to #53300

Some tuple swaps can be compiled to equally efficient IL as using an explicit temp, and others can't. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgWAFAAEDMACFAmDBhDAb2Q1K0xQBYMBBAJzogE8BlAdwgAcBRAD04A2UAMZQALgApYYgNoBdDBAYBKIiTIbpGMQFMwnDAF5FDGQEY5AbnUbSSuuYXH7MgAxWbtl+6Pa9nayQNAF9PT3QsanpGVg5OABUAV0EdKRhZBXtVYiDbUgB6fIwBHQBnUowdADMqkSgddM8NCW85ABoTBwtVYxbTCw7W5UCQ8IpqABkAe2EIAXYuPkE6yS0IDq1gbKayLV19X2ARvIxgXwhjvIhffYDPUNzSMciMadn5uKSUtLFFDfTTttHrZCpUAI6JOYCJiVGp1BpiHakFodLa+CTAQbDe7IYJAA).

We should consider having a setting for this analyzer to only recommend changing to a tuple assignment if the codegen will be equally good. Alternatively, we could use separate diagnostic IDs for when codegen will be equally good versus when extra temps will be introduced.

cc @stephentoub

**Comment by CyrusNajmabadi** [comment]
This is not really an efficiency thing.  This is just in the line of easy-idiom-upgrades.  i.e. the code is nice and clear as to what it does.  

Note: if this is an array, i would expect our codegen to be good here.  I get why the code-gen might be less efficient with other collection-types that we can't peek into.  But with arrays we know the semantics there, and we should be able to emit the efficient version since we know it will have the same user impact, right @RikkiGibson ?

**Comment by stephentoub** [comment]
> This is not really an efficiency thing.

It's an efficiency thing when the IDE recommends you change your code to something less efficient :smile:

**Comment by CyrusNajmabadi** [comment]
Sorry :)   I'm just stating that our "use these idioms" features do not care about perf.  They're just about clarity.  It's similar to our views on not maintaining the same semantics as before.  The view is that more often than not users would prefer the new form, even if perf slightly changes, or semantics has a risk of changing.  It's that reason that these are only suggestions as well.  It's always been a "hey, did you know?" sort of thing. For the users who are in a more restricted camp where perf cannot slip back, or any change in semantics is not ok, we would recommend not using our features.

**Comment by RikkiGibson** [comment]
> Note: if this is an array, i would expect our codegen to be good here.  I get why the code-gen might be less efficient with other collection-types that we can't peek into.  But with arrays we know the semantics there, and we should be able to emit the efficient version since we know it will have the same user impact, right @RikkiGibson ?

Yeah, this looks like a case we could be optimizing but we aren't yet. That's part of what would make a switch on the IDE side tricky. It would have to replicate the logic the compiler uses to decide to optimize.

It is true that if the feature is turned off, it will never make a suggestion which reduces the quality of emitted IL.

**Comment by stephentoub** [comment]
> I'm just stating that our "use these idioms" features do not care about perf

And I'm stating that that's unfortunate ;-)  We're offering suggestions by default to all Visual Studio users that are telling them to do things that are then less efficient.

**Comment by sharwell** [comment]
I think the only way this would work without increasing generated code complexity is if we update to use `Interlocked.Exchange` for the tuple swap pattern here. The writes operations of the swap are reversed relative to the original code, but `Interlocked.Exchange` would ensure the difference is not observable.

**Comment by CyrusNajmabadi** [comment]
Marking for the compiler as it would make sense for this to just get good codegen.  Talking to rikki offline, it def seems possible for a set of cases that are not problematic for the compiler to determine.
